### ISSUE TITLE
How do I debug this?

### ISSUE THREAD

**Issue by dgutov** [castwide/vscode-solargraph#9]
Installed VS Code and the extension to try it out on a real project.

I press `Ctrl-Space`, and the popup just says "Loading..." indefinitely.

The gem is installed, version 0.11.1 and the server is running (started by VS Code).

**Comment by castwide** [comment]
1. What operating system are you using?

2. Does the file you're editing have a .rb extension?

3. What's the code in the file?

4. If you go to Help -> Toggle Developer Tools and open the Console, are there any error messages? Please copy/paste or take a screenshot if you can.

**Comment by dgutov** [comment]
1. Ubuntu 16.10.

2. Yep.

3. It's proprietary 😃 But it's a small-ish Sinatra controller. I'm not getting any completions even at the top-level, before the class definition.

4. No errors, just:

```
[Extension Host] Solargraph extension activated.
extensionHost.ts:335[Extension Host] The Solargraph gem is installed and working.
extensionHost.ts:335[Extension Host] Checking gem version
extensionHost.ts:335[Extension Host] Updating gem yardocs
extensionHost.ts:335[Extension Host] Starting the server
extensionHost.ts:335[Extension Host] [2017-08-29 23:01:56] INFO  WEBrick 1.3.1
[2017-08-29 23:01:56] INFO  ruby 2.3.4 (2017-03-30) [x86_64-linux]

extensionHost.ts:335[Extension Host] == Sinatra (v2.0.0) has taken the stage on 0 for development with backup from WEBrick
[2017-08-29 23:01:56] INFO  WEBrick::HTTPServer#start: pid=19225 port=38409

extensionHost.ts:335[Extension Host] Preparing /home/dgutov/stash/project-name

extensionHost.ts:335[Extension Host] warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.3-compliant syntax, but you are running 2.3.4.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.

extensionHost.ts:335[Extension Host] Getting completion items using workspace /home/dgutov/stash/project-name
extensionHost.ts:335[Extension Host] Solargraph gem version is current
extensionHost.ts:335 [Extension Host] Getting completion items using workspace /home/dgutov/stash/project-name
extensionHost.ts:335 [Extension Host] Preparing /home/dgutov/stash/project-name
extensionHost.ts:335 [Extension Host] 

2extensionHost.ts:335 [Extension Host] Getting completion items using workspace /home/dgutov/stash/project-name
```

(and nothing)

**Comment by dgutov** [comment]
I just had the OS do an OOM kill on a solargraph server process. That could be related.

**Comment by castwide** [comment]
It might be choking on the size of the workspace, but I think it would have to be plenty big. I haven't run into serious issues up to about 10kloc.

One common cause I've encountered is unit tests. I had problems with RSpec suites on a couple of projects. I fixed it by having the gem avoid parsing anything in the spec/ folder by default. If you have tests in a folder with a different name, that might be where it's hanging.

You can try excluding unnecessary directories with a .solargraph.yml file. Let's say, for example, you want to skip a folder named test/. You can do the following:

1. Go to the command palette and run `Create a Solargraph config file`. (This is the same as running `solargraph config` from the command line.) It'll create a .solargraph.yml file with default settings.

2. Add the directory you want to exclude to .solargraph.yml so it looks something like this:

<pre>
    include:
      - ./**/*.rb
    exclude:
      - spec/**/*
      - test/**/*     # <- Your new entry
</pre>

3. Go to the command palette and run `Restart Solargraph`.

(Side note: if you're editing a file in an excluded directory, Solargraph will still analyze that specific file while ignoring everything else.)

I still haven't figured out why test suites in particular are such a drag on performance.

**Comment by dgutov** [comment]
Thanks, that did it. I tried creating such a config previously, but we have a repo with multiple projects in one directory, and now I've opened just one of them as the workspace (and put the config inside it), completion at least does something.

It's very slow, though: after pressing Ctrl-Space, I have to wait like 1-4 seconds, depending on the context, before the menu shows up. It even takes about 1 second in a newly generated Rails project.
